teps/draft-20200911-hermekton.md

```diff
@@ -0,0 +1,222 @@
+---
+title: TEP-NNNN-Hermetic-Builds
+authors:
+  - "@dlorenc"
+creation-date: 2020-09-11
+last-updated: 2020-09-11
+status: proposed
+---
+
+# TEP-NNNN: Hermekton: Hermetic Builds in Tekton Pipelines
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [User Stories (optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Notes/Constraints/Caveats (optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+A [Hermetic Build](https://landing.google.com/sre/sre-book/chapters/release-engineering/)
+is a release engineering best practice for increasing the reliability and consistency of software builds.
+They are self-contained, and do not depend on anything outside of the build environment.
+This means they do not have network access, and cannot fetch dependencies at runtime.
```

Comment:
Do I understand it correctly that in this context, the build environment is a k8s cluster?

Reply:
I can't speak for Dan, but in general we should not assume a Kubernetes environment running the API, even though today and likely a long time in the future Kubernetes will be the underlying platform.

One thing I really like about this proposal is that it _doesn't_ just rely on existing K8s ideas (`NetworkPolicy`, for instance) since this makes it possible to support Hermekton outside of K8s.

Reply:
By "build environment" here I mean the actual filesystem the build process is running in. So in this case it's really just the container image/any injected volumes.